\documentclass{article}
%\documentclass[hyperref]{ctexart}
\usepackage{graphicx}
\usepackage{multirow}
\usepackage{color}
\usepackage{CJK,CJKnumb}
\usepackage[citecolor=black,linkcolor=black,urlcolor=blue,colorlinks=true,CJKbookmarks]{hyperref}


\begin{document}
\begin{CJK*}{GBK}{song}
\renewcommand{\today}{\number\year 年 \number\month 月 \number\day 日}
\renewcommand{\tablename}{表}
\renewcommand{\figurename}{图}
\renewcommand{\abstractname}{摘要}
\renewcommand\refname{参考文献}
\title{第二代高通量测序组装综述}
\author{孙钊\\zixiaojindao@gmail.com}
\maketitle
\newpage

\begin{abstract}
第二代高通量测序产生了大规模的短序列。大规模输入，指数级的序列重叠，高概率的序列重复使得早期基于``overlap-layout-consensus"的方法不再适用。新的基于de bruijn图的方法在短序列组装中取得了成功。本文以一个纯计算机专业的学生的角度阐述了第二代高通量测序方法，序列组装的挑战以及如何进行序列组装的问题。在文末描述了我自己在序列组装领域做出的一些努力。
%Second-generation high-throughput sequencing methods generates large quantity short reads. Challenges including large scale input, exponential overlap among reads, high possibility of repeat sequence make early methods based on ``overlap-layout-consensus" working badly. New methods dependent on de bruijn graph approach succeeds for short reads. This article presents the methods of second generation sequencing, challenges for sequence assembly, how to assemble sequence in a pure computer science student's perspective. At last, my own contributions to sequence assembly are presented.
\end{abstract}
%\section{Before Everything}
\section{开始之前}
首先我想解释一下我写这篇文章的动机。这篇文章的主题是关于基因工程的计算机辅助过程--序列组装。序列组装实际上是生物信息学的范畴。作为一个纯计算机背景的学生，我在2010年的夏天开始在这个领域进行研究。在读了大量的相关领域论文，与国内外的生物科研人员充分交流，尝试了各种各样的生物信息学中的工具之后，我没有取得任何创新性的成果，但是获得了非常宝贵的领域经验，并希望与那些对该领域一无所知又徘徊在领域门口的计算机的学生分享这些经验。本文我讲逐一回答``什么是基因测序",``什么是序列组装以及它的挑战", ``如何组装序列", ``如何评价组装软件"这些问题，以及在最后列举我个人在这个领域做出的一些努力。本文中，我不会涉及太多算法的细节，我主要的目的是给读者一些关于基因序列组转中``做什么"和``如何做"的直觉印象，并且，我会提供大量的相关文章，材料和链接来帮助你构建一个完整的领域框架。
%I would like to first explain the motivation of this article at the beginning. The theme of this article is about sequence assembly which is a computer aided process for constructing gene sequence. It is actually belonging to the scope of bioinformatics. As a pure computer science student, I have started researching in this field since the summer in 2010. After reading lots of papers, discussing with biology researchers from China as well as foreign ones, trying kinds of bioinformatic tools, I approached nothing new but valuable experiences of it. In this article, I will answer the questions of ``what is gene sequencing", ``what is sequence assembly and its challenge", ``how to assembly sequence", ``how to evaluate sequence assembly softwares" and my own contribution to sequence assembly in a computer science researcher's perspective. I'am not going to tell the deep details but giving initiations about what to do and how to do about sequence assembly. Moreover, I will provide valuable references of related papers and materials to help you get a full view of sequence assembly.
%section{What is Gene Sequencing}
\section{什么是基因测序}
在基因工程和生物化学领域，测序意味着确定无支链生物聚合物的一级结构，例如DNA, RNA和蛋白质。本文将除去蛋白质而只关注DNA和RNA的序列分析。DNA测序就是确定一段给定DNA链的核苷酸序列的过程。具体而言，DNA是由4种分别用`A', `G', `C', `T'来表示的核苷酸组成的序列，也就是包含`A'， `G'， `C'， `T'的一个字符串。但是，生物上应用最广泛的鸟枪法测序，随机的把原始的DNA链条打断成碎片DNA，然后对这些碎片DNA进行测序。这些DNA碎片测出来的DNA序列称为一个read。为了提高read的覆盖度和质量，大量的由PCR扩增技术通过细菌作为模板的DNA复制链被同时测序 图\ref{shot_gun_method}显示了鸟枪法测序的过程，同时你会发现，仅碎片DNA的两端被测序了而不是全部。这种只测两端的现象是由生物测序的方法决定的，但是通过加入特别的方法，这一对的read的距离可以被估计出来。这时，这两个read被称作是pair-end 的read，它们之间的距离叫做insert length。\\
%In genetics and biochemistry, sequencing means to determine the primary structure of an unbranched biopolymer. The biopolymer can be DNA, RNA, protein. In this article, I will focus on DNA and RNA sequence analysis excluding protein. DNA sequencing is the process of determining the nucleotide order of a given DNA chain. Concretely, DNA is a chain of four types nucleotide, represented by letter of `A', `G', `C', `T'. DNA sequencing is trying to produce the corresponding string of `A', `G', `C', `T' for a sample DNA chain. However, the most popular biology sequencing method called shot gun, randomly cut the original DNA chain into fragments and a set of `A', `G', `C', `T' strings. Each nucleotide string which is the sequence of a fragment is called a read. To increase the read coverage and read quality, copies of DNA made by PCR amplification with a typical bacteria template are sequenced. Fig. \ref{shot_gun_method} shows the shot gun process, and you may find that reads are sequenced from the two ends of DNA fragments instead of the complete one. The end sequencing phenomena is caused by biology sequencing methods, however, if particular reaction and methods are included, the distance between the two ends can be estimated. In this case, the two end reads are called pair-end reads and the distance is called insert length.\\
\begin{figure}[ht]
  \centering
  % Requires \usepackage{graphicx}
  \includegraphics[width=8cm]{Figure1.jpg}\\
  \caption{}\label{shot_gun_method}
\end{figure}
DNA分子是由两条互补链组成的双螺旋结构。根据碱基配对的原则--`A'配对`T'，`G'配对`C'，可由DNA分子中的一条链推导出互补的那条链。一些用3'和5'表示的特殊序列用来表示两条链的方向。为了简化说明起见，如果一条链定义为3'到5'方向，则其互补链为5'到3'方向。例 1显示了一个双链的碎片DNA，其中红色部分为pair-end的read。注意到pair-end的两条read是分别位于两条互补链上的，但他们测序的方向都是从3' 到5'。因此，例 1中的两条read对应的序列字符串为`AGCTAA' and `GCCAA'。\\
%DNA molecules are double-stranded helices, consisting of two long complement strands. According to base paring principle--`A' complements with `T' and `G' complements with `C', the sequence of a strand can be inferred from its opposite one. Particular sequences denoted by 3' and 5' in the DNA strand is used to specify the orientation of the two strand, and for simplicity, if one strand is specified as 3' to 5', then the opposite one is 5' to 3'. Example 1 shows a double strain DNA fragment with pair-end reads(red string). Note that the pair-end reads are positioned on opposite strand and will be sequenced all from 3' to 5'. So the two pair-end reads string should be `AGCTAA' and `GCCAA'.
\begin{center}
  3'$\rightarrow$5'\\
  {\color{red}AGCTAA}TGCTATCTTGGC\\
  TCGATTACGATAG{\color{red}AACCG}\\
  5'$\rightarrow$3'\\
  例1\\
\end{center}

\noindent 测序方法和平台是随着时间的发展而发展的。第一代测序平台的代表是Sanger\cite{sanger1977nucleotide}, 它测序出来的read平均长度大约为800bp（一般为500-600bp）。最近，新的测序技术已经出现\cite{mardis2008impact}。已经可以商业应用的第二代测序平台有454测序仪\cite{margulies2005genome}，Illu\-mina测序仪\cite{bentley2006whole}和SOLID测序仪(\href{www.appliedbiosystems.com}{www.appliedbiosystems.com})。和传统的Sanger\\
测序平台相比，这些技术节省了大量的花费并且输出的吞吐量很高。但是，第二代测序技术产生的read序列要比传统的Sanger 序列短很多，一般来说454可以达到400-500bp，Illumina为50bp，而SOLID为35bp。正是因为他们长度太短，和早起的测序项目相比，他们必须提高read的覆盖度输出而输出大规模的read序列。\\
%Sequencing methods and platform is developing as time goes. The typical genome analyser of first generation is Sanger\cite{sanger1977nucleotide} producing read lengths of approximately 800bp (typically 500-600bp with non-enriched DNA).Recently, new sequencing methods have emerged \cite{mardis2008impact}. Commercially available technologies include 454 Sequencing \cite{margulies2005genome}, Illumina genome analyser \cite{bentley2006whole} and SOLID sequencing(\href{www.appliedbiosystems.com}{www.appliedbiosystems.com}). Compared to traditional Sanger methods, these technologies function with significantly lower production costs and higher throughput. However, the reads produced by these next-generation sequencing technologies are much shorter than traditional Sanger reads, currently around 400-500 base pairs (bp) for 454, 50bp for Illumina and 35bp for SOLiD. Because of their length, they must be produced in large quantities and at greater coverage depths than earlier sequencing projects.\\
测序芯片把输出的read序列保存在用FASTA或者FASTQ格式的文件中。FAST\-A格式的文件中，每一条read序列由一行描述行开始，然后紧跟着read序列字符串的行。描述行一般都由符号``$>$"开始，read序列建议每行不超过80个字符。图\ref{fasta_format}显示了一个FASTA格式的序列文件。FASTQ文件使用4行来表示一个read序列，第一行一般由`@'符号开始表示一个序列标识符。第二行是原始的read序列字符串。第三行由`+' 号开始一行新的描述标识符。第四行用ascii编码质量数值，其中每个字符的ascii数字对应了第二行同样位置的序列的质量。图\ref{fastq_format}显示了一个最简单的FASTQ文件。如果想了解更多关于质量打分行的细节，可以参考维基百科上的对\href{http://en.wikipedia.org/wiki/FASTQ_format}{FASTQ 格式}的介绍。\\
%Reads are saved as a file by sequencing chip and the most popular format of read file is FASTA and FASTQ. A sequence in FASTA format begins with a single-line description, followed by lines of sequence data. The description line (def-line) is distinguished from the sequence data by a greater-than (``$>$") symbol at the beginning. It is recommended that all lines of text be shorter than 80 characters in length. An example sequence in FASTA format is shown in Fig. \ref{fasta_format}.A FASTQ file normally uses four lines per sequence. Line 1 begins with a '@' character and is followed by a sequence identifier and an optional description (like a FASTA title line). Line 2 is the raw sequence letters. Line 3 begins with a '+' character and is optionally followed by the same sequence identifier (and any description) again. Line 4 encodes the quality values for the sequence in Line 2, and must contain the same number of symbols as letters in the sequence. A minimal FASTQ file might look like this Fig. \ref{fastq_format}. For more details about the quality line, please refer introduction on \href{http://en.wikipedia.org/wiki/FASTQ_format}{FASTQ Format}.\\
最后需要一提的是RNA测序。由于RNA是由DNA转录获得，它所有信息都已经存储在细胞的DNA中。通常对RNA的测序方法首先采取逆转录的反应，将RNA逆转录为cDNA，然后对cDNA采用DNA的测序方法即可。\\
%The last thing to mention is RNA sequencing. As RNA is generated by transcription from DNA, the information is already present in the cell's DNA. The usual method to sequence RNA is first to reverse transcribe the sample to generate cDNA fragments. Then the cDNA are sequenced using DNA sequencing methods.\\
总之第二代测序平台对DNA或RNA进行测序，输出了一个用FASTA或FASTQ文件格式的序列文件。\\
%In a word, the next generation sequencing platform generates a FASTA/FASTQ format file consisting string of description, read and quality(only exists in FASTQ file) for DNA/RNA samples.
\begin{figure}[ht]
  \centering
  % Requires \usepackage{graphicx}
  \includegraphics[width=10cm]{Figure2.jpg}\\
  \caption{}\label{fasta_format}
\end{figure}
\begin{figure}[ht]
  \centering
  % Requires \usepackage{graphicx}
  \includegraphics[width=10cm]{Figure3.jpg}\\
  \caption{}\label{fastq_format}
\end{figure}
%\section{What is sequence assembly}
\section{什么是序列组装}
测序阶段输出的仅仅是一堆碎片DNA的序列，而我们需要的是原始的完整DNA的核苷酸序列。序列组装就要要对齐和合并这些碎片DNA序列为原始序列。直觉上，需要利用read序列之间的重叠区域来把他们粘成一个更长的序列。下面列出的一些挑战使得序列组装成为一个十分复杂的任务：
%What the sequencing stage produces is a collection of DNA fragments's nucleotide sequence, whereas, what we need is the nucleotide sequence of the original one. Sequence assembly tries to align and merge the DNA fragments in order to reconstruct the original sequence. The intuition is to take advantage of overlap information between reads to tie them up and piece into a longer DNA sequence. Challenges listed blow makes sequence assembly a very complicated task.\\
\begin{itemize}
 \item 一个细胞中含有多条染色体，继而含有多条DNA链。这些DNA链同时被测序，不同链上的序列可能会有重叠。
 \item DNA是双链结构，一条链上的序列可能和它的互补链序列重叠。
 \item 随机打断序列的过程并不是一个标准的泊松随机过程，可能会有一部分DNA区域没有被read覆盖到。
 \item DNA序列中可能会有重复出现的序列，这些序列会被错误的当做重叠区域而合并。
 \item 测序错误使得read序列出现插入，删除，突变的字符。
 \item 第二代高通量测序输出较短的read序列，是的read之间重叠的概率增大。
\end{itemize}
%\begin{itemize}
% \item One cell contains multiple chromosomes as well as DNA chains which are sequenced together. DNA fragments from different DNA chain may overlap.
% \item DNA is a double-stranded structure, one strand may overlap with its opposite one.
% \item Sheared fragments along the genome cannot be modeled as a perfect Poisson process so that there may be some region not covered by reads.
% \item DNA sequence may contains repeat regions, and these regions may be incorrectly merged by overlap.
% \item Sequencing errors causes nucleotide deleting, replacing or inserting inside reads.
% \item Second generation sequencing platform produces much shorter reads so that the overlap possibility is exponentially increased.
%\end{itemize}
在上面的这些挑战中，序列重复的问题是最具挑战行的。图\ref{repeat_example}A展示了一个简单的重复序列问题，其中组装软件错误的把两段重复出现的A序列合并成为了一段序列使得最终得到两个连续段而不是一个图\ref{repeat_example}B。\\
%Among those above challenges, the repeat problem is the most challenge thing for short read assembly. A simple example is shown in Fig. \ref{repeat_example}A, where the assembler incorrectly collapses the two copies of repeat A leading to the creation of two contigs instead of one Fig. \ref{repeat_example}B.\\
\begin{figure}[ht]
  \centering
  % Requires \usepackage{graphicx}
  \includegraphics[width=10cm]{Figure4.jpg}\\
  \caption{}\label{repeat_example}
\end{figure}
\noindent 序列组装的基本策略有两种，一种是基于参考序列的，另外一种是全新组装。基于参考序列的方法一般使用对齐工具来把read序列对齐到参考序列上，然后组装出一个和参考序列相似但不完全相同的序列。全新组装策略不使用参考序列，因此对于没有参考序列的物种特别有用。本文仅仅包含对全新组装方法的介绍，并在后续的章节中阐述该方法的细节。\\
%The basic strategies for assembling sequences are categorized by two groups. One is reference based assembly, and the other one is de novo assembly. Reference based methods using a mapping tool to align reads against an existing reference sequence and build a sequence that is similar but not necessarily identical to the reference. The de novo assembly strategy does not use a reference sequence and is much useful for species whose reference sequence are unknown. This article only covers the de novo assembly and will illustrate the details in the subsequent sections.\\
理想情况下，组装程序应该对于每一条测序的染色体输出一个连续段。但实际上，因为上述挑战的存在，组装程序的输出是多个被打断的连续段。总而言之，组装软件输入一个FASTA/FASTQ的read文件，然后输出一个连续段的集合。\\
%Ideally, an assembly program should produce one contig for every chromosome of the genome being sequenced. However, because of challenge3 and challenge4, the output of the assembler is interrupted into a set of contigs. In summary, Inputting a FASTA/FASTQ read file, the assembler outputs a set of contigs.
%\section{How to assembly sequence}
\section{如何组装序列}
第一代测序的序列组装一般是基于``Overlap-consensus-layout"\cite{myers1995toward}方法。 它把每个read看做是一个独立的顶点，如果两个顶点代表的read之间有重叠则给它们之间连一条边。虽然``overlap-consensus-layout"方法是非常直接和有效的，特别是在read很长的情况下。但是对于第二代测序产生的高质量大规模的read序列而言，这种方法用于计算任意一对read之间的重叠部分的代价将是非常大的。历史上仅有一个基于该方法并用来组装短序列的软件--EDENA\cite{hernandez2008novo}。随后，Idury and Waterman提出了一个新的用图的方法来组装短序列，并且这个思想最终发展成为至今为止用来组转短序列的最为流行的方法--De Bruijn图。这个序列的De Bruijn图是以一堆kmer作为顶点，如果其中一个kmer的后k-1个字符串和另外一个kmer的前k-1个字符串相同，那它们之间就连一条边。kmer是用一个固定长度为k 的滑动窗口来切read序列而来的。kmer和de bruijn图的例子可以在图\ref{kmer_example}和图\ref{debruijn_graph_example}中找到。序列de bruijn图有以下性质：
%Overlap-consensus-layout \cite{myers1995toward} is the traditional approach for first generation sequence. It denotes each read as a separate node, where two reads presenting a clean overlap are connected by a bidirected edge. Although overlap-consensus-layout approach is both intuitive and robust, especially in the case of long reads, it is very costly for pair-wise overlap computing when assembling high quantity short reads produced by second generation sequencing platform like SOLID and Illumina. Only one microread assembler, EDENA \cite{hernandez2008novo} was developed using this approach. Later, Idury and Waterman \cite{idury1995new} introduced the use of a sequence graph to represent an assembly, and this idea has been developed into the most popular framework for short read assembly--De Bruijn graph. Sequence De Bruijn graph is constructed from a set of kmers(k is a parameter) as nodes, and two kmers are connected if one kmer's last k-1 nucleotides is identical to the first k-1 nucleotides. kmers are produced by moving a fixed length(k) sliding window on the original read. Examples of kmers and de bruijn graph are show in Fig. \ref{kmer_example} and Fig. \ref{debruijn_graph_example}. The sequence de bruijn graph has following properties:
\begin{itemize}
 \item 长度为n的read产生n-k+1个kmer。
 \item 每一个read唯一对应了de bruijn图中的一条路径。
 \item 每一个顶点通过在最前面或者最后面添加`A'，`G'， `C'， `T'得到它可能的前后各4个邻居。
\end{itemize}
%\begin{itemize}
% \item Read with length of n generates n-k+1 kmers.
% \item Read is mapping to a unique path in the de bruijn graph.
% \item One node has at most 4 successors and 4 predecessors by adding `A', `G', `C', `T' at the first or last of the k-1 nucleotides.
%\end{itemize}
\begin{figure}[ht]
  \centering
  % Requires \usepackage{graphicx}
  \includegraphics[width=10cm]{Figure5.jpg}\\
  \caption{}\label{kmer_example}
\end{figure}
\begin{figure}[ht]
  \centering
  % Requires \usepackage{graphicx}
  \includegraphics[width=10cm]{Figure6.jpg}\\
  \caption{}\label{debruijn_graph_example}
\end{figure}
根据性质2和性质3，de bruijn图的顶点规模的上界为O(n-k+1)，边的上界为O(8(n-k+1)),其中n为目标DNA链的长度。顺便提一下，人类基因组的总长度为3Gbp，它构成的de bruijn图可以存储在一个单独的128G内存的服务器上或者集群系统上。性质3减少了计算任意一对read之间重叠的代价，使得de bruijn图的方法可行。基于这种方法的组装软件有Velvet \cite{zerbino2008velvet}， ABYSS \cite{simpson2009abyss}， ALLPATHS \cite{butler2008allpaths}和SOAPDenovo \cite{li2010novo}。ABYSS和ALLPATHS声称他们组装出来的人类的基因组序列。表\ref{comparison_of_assemblers}展示了这些组装软件的一个简明的对比说明。软件链接可以在表\ref{links_of_assemblers}中找到。\\
%According the property2 and property3, the node scale of the de bruijn graph is bounded by O(n-k+1) and the edge scale of it is bounded by O(8(n-k+1)). By the way, the length of human is about 3G base pairs whose de bruijn graph can be stored in a single server machine or cluster configured with 128GB. Property3 release the heavy cost of searching for all pair-wise overlap, thus making de bruijn graph approach feasible. Softwares include Velvet \cite{zerbino2008velvet}, ABYSS \cite{simpson2009abyss}, ALLPATHS \cite{butler2008allpaths}, SOAPDenovo \cite{li2010novo} follows the de bruijn graph method. ABYSS and ALLPATHS claimed that they successfully assembled human's genome. A brief comparison of these softwares is show in Table \ref{comparison_of_assemblers}. Software links can be found in Table \ref{links_of_assemblers}.\\
我个人阅读过Velvet和ABYSS的源代码，并实际运行过这两个软件，在这里提供它们的一些细节部分。Velvet是由两个模块软件组成的，一个velveth，一个是velvetg，`h'的意思是哈希，`g'的意思是图。Velvet设计了一个新颖的构造de bruijn图的算法并且声称这种方法要比de bruijn图定义的构造方法在单机上更省内存，效率更高。velveth首先顺序的读取read文件，对于每一个read将其拆分为一堆kmer。对于每一个kmer，用一个哈希表来记录它第一次（同时记录它的互补kmer）出现时所在的read的ID（互补kmer的read ID为负）和偏移量。为了保证每个kmer的互补kmer不是自己，k必须取值为奇数。这第一次扫描使得所有的read信息变为了kmer与它前面出现的read之间重叠的信息，这个新的表示结果叫做roadmap，并用一个名为ROADMAP的文件保存。velvetg读取这个ROADMAP文件，计算每个read与它后面的read之间的重叠来产生一个具有相反信息的结构。对于每一个read，一旦它中间有序列与其他read重叠，我们就把它分离出来，那些没有被打断的连续序列就对应一个顶点。根据read之间的重叠关系就可以一个顶点接着顶点地找到它们的邻居。velvetg的最终输出也就是Velvet的最终输出保存为一个名为config.fa文件。ABYSS按照de bruijn图的定义构造图。并行版本（MPI版本）的ABYSS通过一个定义好的哈希函数来确定顶点所属的机器。ABYSS程序由一个用makefile语法写的接口文件``abyss-pe"组织。abyss-pe根据当前的中间文件自动计算出模块的调用依赖，并执行一条最短的可以输出最终结果的调用。ABYSS的最终输出保存为名为*-contigs.fa， *-scanffolds.fa， *-unitigs.fa和一个结果的统计信息文件*-stats。\\
%I've read code and run programs of Velvet and ABYSS and provide some details about them. Velvet consists of two modules named velveth and velvetg where `h' means hash and `g' means graph. Velvet designed a novel graph construction algorithm and claim that it is efficient and memory saved approach on a single machine. velveth sequentially reads the FASTA/FASTQ read file and for each k-mer observed in the set of reads, a hash table records the ID of the first read encountered containing that k-mer and the position of its occurrence within that read. Each k-mer is recorded simultaneously to its reverse complement sequence. To ensure that each k-mer cannot be its own reverse complement, k must be odd. This first scan allows each read to be re-written as a set of new k-mers combined to overlaps with previously hashed reads. This new representation of the reads sequence is called a roadmap which is written into a file named ROADMAP. velvetg reads the ROADMAP file and creats a second database with the opposite information. It records, for each read, which of its original k-mers are overlapped by subsequent reads. The ordered set of original k-mers of that read is cut each time an overlap with another read begins or ends. For each uninterrupted sequence of original k-mers, a node is created. Finally, reads are traced through the graph using the roadmaps. Knowing the correspondence between original k-mers and the newly created nodes, it is possible to proceed from one node to the next, creating a new directed arc or incrementing the multiplicity of an existing one as appropriate at each step. The final output of velvet is produced by velvetg and saved as config.fa. ABYSS constructs de bruijn graph directly following the definition. The MPI version uses a hash function to determine the host machine for a node. The structure of ABYSS program is organized by a make file driver named ``abyss-pe". abyss-pe automatically finds the dependency of abyss modules and perform a minimum number of running modules. The final output is saved as *-contigs.fa, *-scanffolds.fa, *-unitigs.fa and a statistic file named *-stats.\\
\begin{table}[ht]
\begin{center}
\caption{}\label{comparison_of_assemblers}
\begin{tabular}{c|c|c|c}
\hline
软件名称 & 并行性 & 大规模处理 & 内存消耗\\
\hline
Velvet & OpenMP & No & High\\
\hline
ABYSS & OpenMP+MPI & Yes & Low\\
\hline
ALLPATH& ? & Yes & ?\\
\hline
SOAPDenovo & ? & Yes &Low\\
\hline
\end{tabular}
\end{center}
\end{table}
\begin{table}[ht]
\begin{center}
\caption{}\label{links_of_assemblers}
\begin{tabular}{l|l}
\hline
软件名称 &链接\\
\hline
Velvet &\href{http://www.ebi.ac.uk/~zerbino/velvet/}{http://www.ebi.ac.uk/~zerbino/velvet/}\\
\hline
ABYSS &\href{http://www.bcgsc.ca/platform/bioinfo/software/abyss/}{http://www.bcgsc.ca/platform/bioinfo/software/abyss/}\\
\hline
ALLPATH &\href{http://www.broadinstitute.org/science/programs/genome-biology/crd}{http://www.broadinstitute.org/science/programs/genome-biology/crd}\\
\hline
SOAPDenovo &\href{http://soap.genomics.org.cn/soapdenovo.html}{http://soap.genomics.org.cn/soapdenovo.html}\\
\hline
\end{tabular}
\end{center}
\end{table}
\noindent
构造完de bruijn图之后，需要进行的修正图的过程。修正图的思想是通过删除图中的分支路径来把de bruijn图简化为一条长路径。分支路径有两种类型，一种是tips，一种bubbles。直观上讲tips是末端顶点，bubbles是两条共用末端顶点的路径组陈过的``气泡"。一旦一个顶点A唯一指向了一个顶点B，而顶点B也只有一个邻居A指向它，那么这两个顶点就可以合并为一个更长的顶点。Fig .\ref{branch_example}显示了tips和bubbles的样例。需要注意的是重复序列也可以在de bruijn图中产生类似``气泡"的回路，而且我们不知道这些重复的序列在原始序列中重复出现了多少次。在修正图的最后一步，如果输入是pair-end序列的话，我们可以估计回路两端顶点之间距离进而在这两个顶点中找出一条可行的路径。\\
%After building the de bruijn graph, a graph correction process is applied. The idea of the graph correction process is to simplify the original graph into a long path by removing branches. There are two types of branches--tips and bubbles. A tip is a chain of nodes which is disconnected on one end and a bubble is a circle with two common end points. Whenever a node A has only one outgoing arc that points to another node B that has only one incoming arc, the two nodes can be merged into a long node. Tips and bubbles are shown in Fig .\ref{branch_example}. Note that repeat region also creates circles in the graph, and we can not determine how many times should the repeat sequence occur. In the last step, if pair-end reads are provided, we can estimate the distance of seed nodes and find a feasible path between them.\\
最后需要提的是RNA序列组装。RNA序列组装要比DNA序列组装更为复杂，主要原因如下：
%The last thing need to mention is that RNA assembly also named transcript assembly. RNA assembly is more complex than DNA assembly, observing the reasons:
\begin{itemize}
 \item 一些转录序列具有低表达水平，而另外一些具有高表达水平。
 \item 出现测序错误的高表达水平的序列可能比测序正确的低表达水平的序列多很多。
 \item 不同的转录组序列可能从DNA的同一段序列编码，因此有可能被错误的合并为同一个转录序列。
\end{itemize}
%\begin{itemize}
% \item some transcripts have low coverage, whereas others are highly expressed.
% \item reads with sequencing errors derived from a highly expressed transcript may be more abundant than correct reads from a transcript that is not highly expressed.
% \item transcripts encoded by adjacent loci can overlap and thus can be erroneously fused to form a chimeric transcript.
%\end{itemize}
直接使用DNA组装软件去组装RNA序列只会得到更坏的结果。Velvet和ABYSS\\
已经把它们组装DNA的软件扩展为相应的组装RNA的软件，分别为Velvet\_Oases \cite{schulz2012oases}和Trans-ABYSS \cite{robertson2010novo}。最近，一个专门为RNA组装设计的软件Trinity \cite{grabherr2011full}被开发出来。Zhao等人\cite{zhao2011optimizing}对上述三个软件的性能，结果和资源消耗做了全面的比较，得出了Trinity组装出来的结果要比Velvet\_Oasese和Trans-ABYSS好，但是时间代价和内存代价非常大。随后一组专门做HPC的研究人员\cite{henschel2012trinity}对Trinity进行了全面的分析和调优，大大的降低了Trinity的时间和资源代价，使得Trinity成为组装RNA的首选软件。
%Directly using DNA assemblers to assemble RNA sequences only outputs rather bad results. Velvet and ABYSS have extend themselves into Velvet\_Oases \cite{schulz2012oases} and Trans-ABYSS \cite{robertson2010novo} separately to address the additional challenge presented by RNA assembly. Recently, Trinity \cite{grabherr2011full} was specially designed for transcript assembly. Zhao et. al \cite{zhao2011optimizing} made a early comparison of transcript assembly and give a full view of performance, result, resource usage evaluation. It shows Trinity has a relative better assembling ability than Velvet\_Oases and Trans-ABYSS but much slower and memory cost while assembling large scale sequences. A group of HPC developers \cite{henschel2012trinity} made efforts to accelerating trinity and reducing its memory usage which make trinity a first choice when assembling transcript sequences.
\begin{figure}[ht]
  \centering
  % Requires \usepackage{graphicx}
  \includegraphics[width=10cm]{Figure7.jpg}\\
  \caption{}\label{branch_example}
\end{figure}
\section{如何评测组装软件}
%\section{How to Evaluate Assembler}
为了评测现有的组装软件或者我们自己设计的组装软件，第一件需要做的事情就是准备read序列数据。通常有两种数据集可以用来评测，一种是人工合成的模拟数据，一种是真实测序数据。dwgsim \footnote{http://sourceforge.net/apps/mediawiki/dnaa/index.php?title=Whole\_Genome\_Simulation}是基于参考序列的一个全基因组模拟测序软件。参考序列可以在\href{ftp://ftp.ncbi.nlm.nih.gov/refseq/}{ftp://ftp.ncbi.nlm.nih.gov/refseq/}, \href{ftp://ftp.ncbi.nlm.nih.gov/genomes/}{ftp://ftp.ncbi.nlm.nih.gov/genomes/}或者 NCBI's nucleotide数据库\\(\href{http://www.ncbi.nlm.nih.gov/nucleotide}{http://www.ncbi.nlm.nih.gov/nucleotide})中找到。关于refseq的介绍你可以参考\href{http://www.ncbi.nlm.nih.gov/books/NBK50679/}{http://www.ncbi.nlm.nih.gov/books/NBK50679/}。NCBI的SRA数据库\footnote{http://www.ncbi.nlm.nih.gov/sra}提供了大量的真实测序数据。用户可以通过以SRP（研究），SRS（样本），SRX（\\
实验），SRR（实例）开头的访问码来检索这个数据库。同一个样本的多个SRR数据可以用在一起来组装序列。顺便提一下，从SRA数据库下载的数据是以sra格式保存的，用户需要使用\href{http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=software}{NCBI SRA Toolkit}中的fastq-dump来抽取出FASTA/FASTQ格式的文件。如果数据是pair-end类型，需要使用fastq-dump的`--split-3' 或者 `--split-files'参数。评价测序结果的度量有N50长度，运行时间，资源消耗，95\%对齐率等。给定一个连续段的集合（也就是组装软件的输出），N50长度被定义为L，对于所有长度大于L的连续段，它们长度的总和恰好超过所有连续段总和的50\%。N50长度越长就越说明结果中长连续段多，一般也就认为结果越好。在linux上，记录软件运行的时间，内存代价可以使用`/usr/bin/time -v'命令来记录。95\% 对齐率定义为连续段可以以95\%的相似度对齐到参考序列上的数量除以总的连续段的数量。BLAST \cite{altschul1990basic}和BLAT \cite{kent2002blat}是两个最著名的序列对齐工具。由于BLAT效率更高，我一般选择BLAT作为对齐工具。BLAT输出的是一张用PSL格式存储的大表格，其中每一行代表了一个查询序列对齐到参考序列上的信息。对于PSL格式和表格中每一列含义不熟悉的用户可以参考\href{http://genome.ucsc.edu/FAQ/FAQformat.html#format2}{PSL格式}。每一个查询序列可能有多个对齐位置，也就是在PSL文件中有连续多行对齐信息，BLAT 软件没有提供像在web BLAT\footnote{http://www.genoscope.cns.fr/blat-server/cgi-bin/vitis/webBlat}上的相似度和得分两列，需要用户自己计算。计算的方法可以参考\href{http://www.genoscope.cns.fr/blat-server/cgi-bin/vitis/webBlat}{http://www.genoscope.cns.fr/blat-server/cgi-bin/vitis/webBlat}，然后使用最佳匹配来计算95\%对齐率。
%To evaluate present assemblers or assemblers designed by ourselves, the first thing to do is to prepare input data. Usually, two kinds of data are used to evaluate assemblers, synthetic data and real data. dwgsim \footnote{http://sourceforge.net/apps/mediawiki/dnaa/index.php?title=Whole\_Genome\_Simulation} is whole genome simulator based on a reference genome. Reference genome can be found on \href{ftp://ftp.ncbi.nlm.nih.gov/refseq/}{ftp://ftp.ncbi.nlm.nih.gov/refseq/}, \href{ftp://ftp.ncbi.nlm.nih.gov/genomes/}{ftp://ftp.ncbi.nlm.nih.gov/genomes/} or NCBI's nucleotide database (\href{http://www.ncbi.nlm.nih.gov/nucleotide}{http://www.ncbi.nlm.nih.gov/nucleotide}). Details about refseq is explained on \href{http://www.ncbi.nlm.nih.gov/books/NBK50679/}{http://www.ncbi.nlm.nih.gov/books/NBK50679/}. NCBI's SRA database \footnote{http://www.ncbi.nlm.nih.gov/sra} provides a large collection of read read sequences. Users search this database using an access code begin with SRP(studies), SRS(samples), SRX(experients) or SRR(runs). Multiple SRRs of same samples can be used together as input. By the way, files downloaded from SRA database are in sra format and users should dump FASTA/FASTQ format using fastq-dump from \href{http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=software}{NCBI SRA Toolkit}. If pair-end data are expected to be dumped, pass `--split-3' or `--split-files' to fastq-dump. Metrics for evaluate assembler can be N50 length, running time, memory cost, 95\% map ratio and so on. Given a set of contigs(output of assemblers) of varying lengths, the N50 length is defined as the length N for which half of all bases in the sequences are in a sequence of length L $<$ N. Obviously, the larger the N50 length, the better the assembled contigs. On linux like systems, users can use `/usr/bin/time -v' command to record the running time and peak memory usage of a user process. The 95\% mapping ratio is defined as the number of contigs with at least 95\% identity mapping to the reference dividing by the total number of contigs. BLAST \cite{altschul1990basic} and BLAT \cite{kent2002blat} are two popular mapping tools for aligning contigs to references. I often choose BLAT for reasons of efficiency. BLAT outputs a very big table(PSL format) to store the mapping info for each query sequence. Users who are strange to PSL format and the meaning of each column should refer to \href{http://genome.ucsc.edu/FAQ/FAQformat.html#format2}{PSL format}. Each query may have multiple candidate alignment positions, and BLAT program doesn't provide a percent identity value column or score column like the web BLAT \footnote{http://www.genoscope.cns.fr/blat-server/cgi-bin/vitis/webBlat}. Users can replicate the two columns referring to \href{http://www.genoscope.cns.fr/blat-server/cgi-bin/vitis/webBlat}{http://www.genoscope.cns.fr/blat-server/cgi-bin/vitis/webBlat} and use the best hit to calculate 95\% mapping ratio.
\section{个人贡献}
%\section{My Own Contribution}
我个人对序列组装的贡献列举在表\ref{my_own_contribution}中。
%My own contribution to the sequence assembly is shown in Table \ref{my_own_contribution}.
\begin{table}[ht]
\begin{center}
\caption{}\label{my_own_contribution}
\begin{tabular}{l|l}
\hline
贡献 & 链接\\
\hline
Velvet and Oases for windows&\href{https://github.com/zixiaojindao/velvet.git}{https://github.com/zixiaojindao/velvet.git}\\
\hline
ABYSS for windows &\href{https://github.com/zixiaojindao/ABYSS.git}{https://github.com/zixiaojindao/ABYSS.git}\\
\hline
MemoryUsageMonitor for windows &\href{https://github.com/zixiaojindao/MemoryUsageMonitor.git}{https://github.com/zixiaojindao/MemoryUsageMonitor.git}\\
\hline
blat-statistics for windows &\href{https://github.com/zixiaojindao/blat-statistics.git}{https://github.com/zixiaojindao/blat-statistics.git}\\
\hline
Sequence Assembly Information & \href{https://github.com/zixiaojindao/RNA-sequence-info.git}{https://github.com/zixiaojindao/RNA-sequence-info.git}\\
\hline
\end{tabular}
\end{center}
\end{table}
\section{总结}
%\section{Summary}
从一个程序员的角度看，序列组装就是通过序列之间的重叠把碎片序列的字符串粘贴成为较长原始字符串的过程。但是实际上，因为各种生物相关的挑战存在，使得组装任务变得非常复杂。大部分现有的基于de bruijn图的方法取得了较为满意结果。但是组装软件的时间和资源代价在组装大规模序列时依然十分可观。希望新的计算机算法和生物技术的发展可以在未来帮助尽快解决这一问题。
%In a programmer's perspective, sequence assembly is piecing string fragments into a set of long strings using overlap among them. However, things are much complicated in practice due to biological challenges. Most of assembly softwares follow the de bruijn graph approach and reach a satisfying point. However, the running time and memory cost for large scale sequence is still not very good as well as assembled contigs. Hope new computer algorithms and biological methods could help address this problem much better in the future.
\bibliographystyle{plain}
\bibliography{Thesis}
\end{CJK*}
\end{document}
